.github/workflows/ci.yml

```diff
@@ -5,24 +5,98 @@ jobs:
+    - name: Set common variables
+      run: |
+        echo "BASE_IMAGE_NAME=nextstrain/base" >> $GITHUB_ENV
+        echo "BUILDER_IMAGE_NAME=nextstrain/base-builder" >> $GITHUB_ENV
+        echo "TAG_AND_PUSH=${{ toJSON(github.event_name != 'pull_request') }}" >> $GITHUB_ENV
+        echo "ON_DEFAULT_BRANCH=${{ toJSON(github.ref_type == 'branch' && github.ref_name == github.event.repository.default_branch) }}" >> $GITHUB_ENV
+
+    - if: ${{ fromJSON(env.TAG_AND_PUSH) && fromJSON(env.ON_DEFAULT_BRANCH) }}
+      name: Set tag name (default branch)
+      run: echo "TAG_NAME=build-$CACHE_DATE" >> $GITHUB_ENV
+
+    # From `man docker-image-tag`: The tag name must be valid ASCII and may
+    # contain lowercase and uppercase letters, digits, underscores, periods
+    # and hyphens.
+    # TODO: parameterize this as workflow input
+    - if: ${{ fromJSON(env.TAG_AND_PUSH) && !fromJSON(env.ON_DEFAULT_BRANCH) }}
+      name: Set tag name (non-default branch)
+      run: echo "TAG_NAME=branch-${GITHUB_REF_NAME//[^A-Za-z0-9._-]/-}" >> $GITHUB_ENV
+
+    - if: ${{ fromJSON(env.TAG_AND_PUSH) }}
+      name: Set tags to push
+      run: |
+        maybe_base_image_latest_tag=${{ fromJSON(env.ON_DEFAULT_BRANCH) && ',$BASE_IMAGE_NAME:latest' || '' }}
+        echo "BASE_IMAGE_TAGS=$BASE_IMAGE_NAME:$TAG_NAME$maybe_base_image_latest_tag" >> $GITHUB_ENV
+        maybe_builder_image_latest_tag=${{ fromJSON(env.ON_DEFAULT_BRANCH) && ',$BUILDER_IMAGE_NAME:latest' || '' }}
+        echo "BUILDER_IMAGE_TAGS=$BUILDER_IMAGE_NAME:$TAG_NAME$maybe_builder_image_latest_tag" >> $GITHUB_ENV
+
+    - if: ${{ fromJSON(env.TAG_AND_PUSH) }}
       uses: docker/login-action@v1
       with:
         username: ${{ secrets.DOCKER_LOGIN }}
         password: ${{ secrets.DOCKER_PASSWORD }}
-    - if: ${{ github.event_name != 'pull_request' && github.ref_type == 'branch' && github.ref_name == github.event.repository.default_branch }}
-      run: |
-        ./devel/tag build-$CACHE_DATE
-        ./devel/push latest build-$CACHE_DATE
-    - if: ${{ github.event_name != 'pull_request' && github.ref_type == 'branch' && github.ref_name != github.event.repository.default_branch }}
-      # From `man docker-image-tag`: The tag name must be valid ASCII and may
-      # contain lowercase and uppercase letters, digits, underscores, periods
-      # and hyphens.
-      run: |
-        tag=branch-${GITHUB_REF_NAME//[^A-Za-z0-9._-]/-}
-        ./devel/tag $tag
-        ./devel/push $tag
+
+    # The nextstrain/base Dockerfile is a multi-stage with both a "builder" target
+    # and a main target.  To enable proper caching via --cache-from we need both
+    # these images available to pull layers from.  This means pulling both in at
+    # the start and pushing both up at the end.
+
+    # Calling `docker run nextstrain/base` will still only pull down the small base
+    # image rather than pulling down the larger nextstrain/base-builder image.
```

Comment:
I believe one of the features of buildx is better support for multi-stage builds, such that you can do just `build` operation for a multi-stage build but still push up tagged images for each stage (instead of having to explicitly `build` each stage yourself just so you can push each later).

I'm not sure how building both stages in parallel would work since the second stage relies on the first?

Reply:
Ah, this better support appears to be ([1](https://github.com/docker/buildx/blob/master/docs/reference/buildx_build.md#-export-build-cache-to-an-external-cache-destination---cache-to), [2](https://github.com/moby/buildkit#export-cache)) in the form of exporting cached layers via `type=registry` with `mode=max` (meaning "all intermediate layers"), but you have to use a separate image tag/name (unlike `type=inline`, but that only supports `mode=min` which means "final layers").